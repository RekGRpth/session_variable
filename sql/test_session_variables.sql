/*
 * Copyright (c) Splendid Data Product Development B.V. 2013
 *
 * This program is free software: You may redistribute and/or modify under the
 * terms of the GNU General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at Client's option) any
 * later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
 * details.
 *
 * You should have received a copy of the GNU General Public License along with
 * this program.  If not, Client should obtain one via www.gnu.org/licenses/.
 */

-- set client_min_messages = DEBUG4;

-- init
create extension session_variable;
select session_variable.init();

select session_variable.get_session_variable_version();

select session_variable.get('does not exist', null::text);                                 -- fails: no variables exist

select session_variable.create_variable('some_date', 'date'::regtype, '2015-07-15'::date);
select session_variable.create_variable('an integer', 'integer'::regtype, 123456789);
select session_variable.create_variable('integer_initially_zero', 'integer'::regtype, 0);
select session_variable.create_variable('just text', 'text'::regtype, 'just some text'::text);
select session_variable.create_variable('varchar', 'varchar'::regtype, 'a varchar text'::varchar);
select session_variable.create_constant('numeric const', 'numeric'::regtype, 123.45);
select session_variable.create_variable('initially null', 'text'::regtype, null::text);
select session_variable.create_variable('a_record', 'record'::regtype, ('some text', '2015-07-17'::date, 123.45::numeric));  -- fails: pseudo type
select session_variable.create_variable('int[][]', 'int[][]'::regtype, '{{1,2},{3,4},{5,6}}'::int[][]);

select session_variable.create_variable('wrongdate', 'interval'::regtype, '2015-07-15'::date); -- fails: wrong datatype
select session_variable.get('does not exist', null::text);                                 -- fails: does not exist

select session_variable.type_of('some_date');
select session_variable.type_of('an integer');
select session_variable.type_of('varchar');
select session_variable.type_of('numeric const');
select session_variable.type_of('int[][]');
select session_variable.type_of('does not exist');                                         -- fails: does not exist
select session_variable.type_of(null);                                                     -- fails: variable_or_constan_name must be filled

select to_char(session_variable.get('some_date', null::date), 'yyyy-mm-dd');
select session_variable.get('an integer', null::integer);
select session_variable.get('just text', null::text);
select session_variable.get('varchar', null::varchar);
select session_variable.get('numeric const', null::numeric);
select session_variable.get('numeric const', null::date);                                  -- fails: wrong type
select session_variable.get('int[][]', null::int[][]);
select session_variable.get('initially null', null::text);
select session_variable.get('a_record', null::record);                                     -- fails: not created because of pseudo type


select session_variable.exists('int[][]');
select session_variable.exists('initially null');
select session_variable.exists('does not exist');

select session_variable.drop('int[][]');

select session_variable.set('some_date', '2050-12-11'::date);
select session_variable.set('an integer', 987654321);
select session_variable.set('just text', 'a new bit of text'::text);
select session_variable.set('varchar', 'an altered varchar text'::varchar);
select session_variable.set('initially null', 'now filled'::text);

select session_variable.set('numeric const', 223344);                                      -- fails: constants cannot be set
select session_variable.set('does not exist', null::text);                                 -- fails: does not exist

select to_char(session_variable.get('some_date', null::date), 'yyyy-mm-dd');
select session_variable.get('an integer', null::integer);
select session_variable.get('just text', null::text);
select session_variable.get('varchar', null::varchar);
select session_variable.get('numeric const', null::numeric);
select session_variable.set('initially null', null::text);

select session_variable.init();

select to_char(session_variable.get('some_date', null::date), 'yyyy-mm-dd');
select session_variable.get('an integer', null::integer);
select session_variable.get('integer_initially_zero', null::integer);
select session_variable.get('just text', null::text);
select session_variable.get('varchar', null::varchar);
select session_variable.set('initially null', null::text);
select session_variable.get('numeric const', null::numeric);

select session_variable.alter_value('some_date', '1900-01-01'::date);
select session_variable.alter_value('an integer', 0);
select session_variable.alter_value('just text', ''::text);
select session_variable.alter_value('varchar', null::varchar);
select session_variable.alter_value('numeric const', 3333333333333333333333.3333333333333);

select session_variable.alter_value('does not exist', null::text);                         -- fails: does not exist
select session_variable.alter_value('an integer', null::date);                             -- fails: wrong datatype

select to_char(session_variable.get('some_date', null::date), 'yyyy-mm-dd');
select session_variable.get('an integer', null::integer);
select session_variable.get('just text', null::text);
select session_variable.get('varchar', null::varchar);
select session_variable.get('numeric const', null::numeric);

select session_variable.set('an integer', 123.456::float);
select session_variable.get('an integer', null::smallint);

select session_variable.set('varchar', '2018-03-15'::date);
select session_variable.get('varchar', null::char);
select session_variable.set('varchar', '12345'::text);
select session_variable.set('varchar', 23.456::numeric);                                   -- fails: the result cannot be cast from varchar to numeric

select session_variable.drop('some_date');
select session_variable.drop('an integer');

select session_variable.dump();
select session_variable.dump(false);

select to_char(session_variable.get('some_date', null::date), 'yyyy-mm-dd');               -- fails: has just been dropped
select session_variable.get('an integer', null::integer);                                  -- fails: has just been dropped
select session_variable.get('just text', null::text);
select session_variable.get('varchar', null::varchar);
select session_variable.get('numeric const', null::numeric);

create or replace function session_variable.variable_initialisation()
   returns void language plpgsql as $$
begin
	if session_variable.is_executing_variable_initialisation()
	then
	    raise notice 'executing session_variable.variable_initialisation() on behalf of session_variable initialisation code';
	    select session_variable.set('numeric const', 223344);                              -- setting a constant should succeed now  
	else
	    raise notice 'executing session_variable.variable_initialisation() on its own account';
	end if;
end;
$$; 
select session_variable.get('numeric const', null::numeric);
select session_variable.variable_initialisation();
select session_variable.get('numeric const', null::numeric);
select session_variable.init();
select session_variable.get('numeric const', null::numeric);

drop function session_variable.variable_initialisation();
select session_variable.init();
select session_variable.get('numeric const', null::numeric);

select session_variable.drop('just text');
select session_variable.drop('varchar');
select session_variable.drop('numeric const');
select session_variable.drop('initially null');
select session_variable.drop('does not exist');                                            -- fails: does not exist

select to_char(session_variable.get('some_date', null::date), 'yyyy-mm-dd');               -- fails: has just been dropped
select session_variable.get('an integer', null::integer);                                  -- fails: has just been dropped
select session_variable.get('just text', null::text);                                      -- fails: has just been dropped
select session_variable.get('varchar', null::varchar);                                     -- fails: has just been dropped
select session_variable.get('numeric const', null::numeric);                               -- fails: has just been dropped

-- cleanup
drop schema if exists session_variable cascade;