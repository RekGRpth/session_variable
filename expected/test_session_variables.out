/*
 * Copyright (c) Splendid Data Product Development B.V. 2013
 *
 * This program is free software: You may redistribute and/or modify under the
 * terms of the GNU General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at Client's option) any
 * later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
 * details.
 *
 * You should have received a copy of the GNU General Public License along with
 * this program.  If not, Client should obtain one via www.gnu.org/licenses/.
 */
-- set client_min_messages = DEBUG4;
-- init
create extension session_variable;
select session_variable.init();
 init 
------
    0
(1 row)

select session_variable.get_session_variable_version();
 get_session_variable_version 
------------------------------
 3.2
(1 row)

select session_variable.get('does not exist', null::text);                                 -- fails: no variables exist
ERROR:  variable or constant 'does not exist' does not exists
select session_variable.create_variable('some_date', 'date'::regtype, '2015-07-15'::date);
 create_variable 
-----------------
 t
(1 row)

select session_variable.create_variable('some_name', 'name'::regtype, 'a_table_or_view_name'::name);
 create_variable 
-----------------
 t
(1 row)

select session_variable.create_variable('an integer', 'integer'::regtype, 123456789);
 create_variable 
-----------------
 t
(1 row)

select session_variable.create_variable('integer_initially_zero', 'integer'::regtype, 0);
 create_variable 
-----------------
 t
(1 row)

select session_variable.create_variable('just text', 'text'::regtype, 'just some text'::text);
 create_variable 
-----------------
 t
(1 row)

select session_variable.create_variable('varchar', 'varchar'::regtype, 'a varchar text'::varchar);
 create_variable 
-----------------
 t
(1 row)

select session_variable.create_constant('numeric const', 'numeric'::regtype, 123.45);
 create_constant 
-----------------
 t
(1 row)

select session_variable.create_variable('initially null', 'text'::regtype, null::text);
 create_variable 
-----------------
 t
(1 row)

select session_variable.create_variable('a_record', 'record'::regtype, ('some text', '2015-07-17'::date, 123.45::numeric));  -- fails: pseudo type
ERROR:  pseudo types are not supported: record
select session_variable.create_variable('int[][]', 'int[][]'::regtype, '{{1,2},{3,4},{5,6}}'::int[][]);
 create_variable 
-----------------
 t
(1 row)

select session_variable.create_variable('wrongdate', 'interval'::regtype, '2015-07-15'::date); -- fails: wrong datatype
ERROR:  value must be of type date, but is of type interval
select session_variable.get('does not exist', null::text);                                 -- fails: does not exist
ERROR:  variable or constant 'does not exist' does not exists
select session_variable.type_of('some_date');
 type_of 
---------
 date
(1 row)

select session_variable.type_of('some_name');
 type_of 
---------
 name
(1 row)

select session_variable.type_of('an integer');
 type_of 
---------
 integer
(1 row)

select session_variable.type_of('varchar');
      type_of      
-------------------
 character varying
(1 row)

select session_variable.type_of('numeric const');
 type_of 
---------
 numeric
(1 row)

select session_variable.type_of('int[][]');
  type_of  
-----------
 integer[]
(1 row)

select session_variable.type_of('does not exist');                                         -- fails: does not exist
ERROR:  variable or constant 'does not exist' does not exists
select session_variable.type_of(null);                                                     -- fails: variable_or_constan_name must be filled
ERROR:  variable or constant name must be filled
select to_char(session_variable.get('some_date', null::date), 'yyyy-mm-dd');
  to_char   
------------
 2015-07-15
(1 row)

select session_variable.get('some_name', null::name);
         get          
----------------------
 a_table_or_view_name
(1 row)

select session_variable.get('an integer', null::integer);
    get    
-----------
 123456789
(1 row)

select session_variable.get('just text', null::text);
      get       
----------------
 just some text
(1 row)

select session_variable.get('varchar', null::varchar);
      get       
----------------
 a varchar text
(1 row)

select session_variable.get('numeric const', null::numeric);
  get   
--------
 123.45
(1 row)

select session_variable.get('numeric const', null::date);                                  -- fails: wrong type
ERROR:  The variable's internal type numeric, cannot be cast to type date
select session_variable.get('int[][]', null::int[][]);
         get         
---------------------
 {{1,2},{3,4},{5,6}}
(1 row)

select session_variable.get('initially null', null::text);
 get 
-----
 
(1 row)

select session_variable.get('a_record', null::record);                                     -- fails: not created because of pseudo type
ERROR:  variable or constant 'a_record' does not exists
select to_char(session_variable.get_stable('an integer', null::real), '9D99999EEEE');
   to_char    
--------------
  1.23457e+08
(1 row)

select session_variable.get_stable('numeric const', null::numeric);
 get_stable 
------------
     123.45
(1 row)

select session_variable.get_constant('numeric const', null::real);
 get_constant 
--------------
       123.45
(1 row)

select session_variable.get_constant('an integer', null::bigint);                          -- fails because variable 'an integer' is not a constant
ERROR:  'an integer' is not a constant
select session_variable.exists('int[][]');
 exists 
--------
 t
(1 row)

select session_variable.exists('initially null');
 exists 
--------
 t
(1 row)

select session_variable.exists('does not exist');
 exists 
--------
 f
(1 row)

select session_variable.drop('int[][]');
 drop 
------
 t
(1 row)

select session_variable.set('some_date', '2050-12-11'::date);
 set 
-----
 t
(1 row)

select session_variable.set('some_name', 'another_name'::varchar);
 set 
-----
 t
(1 row)

select session_variable.set('an integer', 987654321);
 set 
-----
 t
(1 row)

select session_variable.set('just text', 'a new bit of text'::text);
 set 
-----
 t
(1 row)

select session_variable.set('varchar', 'an altered varchar text'::varchar);
 set 
-----
 t
(1 row)

select session_variable.set('initially null', 'now filled'::text);
 set 
-----
 t
(1 row)

select session_variable.set('numeric const', 223344);                                      -- fails: constants cannot be set
ERROR:  constant "numeric const" cannot be set
select session_variable.set('does not exist', null::text);                                 -- fails: does not exist
ERROR:  variable "does not exist" does not exists
select to_char(session_variable.get('some_date', null::date), 'yyyy-mm-dd');
  to_char   
------------
 2050-12-11
(1 row)

select session_variable.get('an integer', null::integer);
    get    
-----------
 987654321
(1 row)

select session_variable.get('some_name', null::text);
     get      
--------------
 another_name
(1 row)

select session_variable.get('just text', null::text);
        get        
-------------------
 a new bit of text
(1 row)

select session_variable.get('varchar', null::varchar);
           get           
-------------------------
 an altered varchar text
(1 row)

select session_variable.get('numeric const', null::numeric);
  get   
--------
 123.45
(1 row)

select session_variable.set('initially null', null::text);
 set 
-----
 t
(1 row)

select session_variable.set('some_name', null::name);
 set 
-----
 t
(1 row)

select coalesce(session_variable.get('some_name', null::text), 'is null');
 coalesce 
----------
 is null
(1 row)

select session_variable.init();
 init 
------
    8
(1 row)

select to_char(session_variable.get('some_date', null::date), 'yyyy-mm-dd');
  to_char   
------------
 2015-07-15
(1 row)

select session_variable.get('an integer', null::integer);
    get    
-----------
 123456789
(1 row)

select session_variable.get('integer_initially_zero', null::integer);
 get 
-----
   0
(1 row)

select session_variable.get('just text', null::text);
      get       
----------------
 just some text
(1 row)

select session_variable.get('varchar', null::varchar);
      get       
----------------
 a varchar text
(1 row)

select session_variable.set('initially null', null::text);
 set 
-----
 t
(1 row)

select session_variable.get('numeric const', null::numeric);
  get   
--------
 123.45
(1 row)

select session_variable.alter_value('some_date', '1900-01-01'::date);
 alter_value 
-------------
 t
(1 row)

select session_variable.alter_value('an integer', 0);
 alter_value 
-------------
 t
(1 row)

select session_variable.alter_value('just text', ''::text);
 alter_value 
-------------
 t
(1 row)

select session_variable.alter_value('varchar', null::varchar);
 alter_value 
-------------
 t
(1 row)

select session_variable.alter_value('numeric const', 3333333333333333333333.3333333333333);
 alter_value 
-------------
 t
(1 row)

select session_variable.alter_value('does not exist', null::text);                         -- fails: does not exist
ERROR:  Variable "does not exist" does not exists
select session_variable.alter_value('an integer', null::date);                             -- fails: wrong datatype
 alter_value 
-------------
 t
(1 row)

select to_char(session_variable.get('some_date', null::date), 'yyyy-mm-dd');
  to_char   
------------
 1900-01-01
(1 row)

select session_variable.get('an integer', null::integer);
 get 
-----
    
(1 row)

select session_variable.get('just text', null::text);
 get 
-----
 
(1 row)

select session_variable.get('varchar', null::varchar);
 get 
-----
 
(1 row)

select session_variable.get('numeric const', null::numeric);
                 get                  
--------------------------------------
 3333333333333333333333.3333333333333
(1 row)

select session_variable.set('an integer', 123.456::float);
 set 
-----
 t
(1 row)

select session_variable.get('an integer', null::smallint);
 get 
-----
 123
(1 row)

select session_variable.set('varchar', '2018-03-15'::date);
 set 
-----
 t
(1 row)

select session_variable.get('varchar', null::char);
    get     
------------
 03-15-2018
(1 row)

select session_variable.set('varchar', '12345'::text);
 set 
-----
 t
(1 row)

select session_variable.set('varchar', 23.456::numeric);                                   -- fails: the result cannot be cast from varchar to numeric
 set 
-----
 t
(1 row)

select session_variable.drop('some_date');
 drop 
------
 t
(1 row)

select session_variable.drop('an integer');
 drop 
------
 t
(1 row)

select session_variable.dump();
                                                                         dump                                                                         
------------------------------------------------------------------------------------------------------------------------------------------------------
 truncate table session_variable.variables;
 select session_variable.init();
 select session_variable.create_variable('initially null', 'pg_catalog.text'::regtype, NULL::pg_catalog.text);
 select session_variable.create_variable('integer_initially_zero', 'pg_catalog.int4'::regtype, '0'::pg_catalog.int4);
 select session_variable.create_variable('just text', 'pg_catalog.text'::regtype, ''::pg_catalog.text);
 select session_variable.create_constant('numeric const', 'pg_catalog.numeric'::regtype, '3333333333333333333333.3333333333333'::pg_catalog.numeric);
 select session_variable.create_variable('some_name', 'pg_catalog.name'::regtype, 'a_table_or_view_name'::pg_catalog.name);
 select session_variable.create_variable('varchar', 'pg_catalog.varchar'::regtype, NULL::pg_catalog.varchar);
(8 rows)

select session_variable.dump(false);
                                                                                                  dump                                                                                                   
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 select session_variable.init();
 select session_variable.create_variable('initially null', 'pg_catalog.text'::regtype, NULL::pg_catalog.text) where not session_variable.exists('initially null');
 select session_variable.create_variable('integer_initially_zero', 'pg_catalog.int4'::regtype, '0'::pg_catalog.int4) where not session_variable.exists('integer_initially_zero');
 select session_variable.create_variable('just text', 'pg_catalog.text'::regtype, ''::pg_catalog.text) where not session_variable.exists('just text');
 select session_variable.create_constant('numeric const', 'pg_catalog.numeric'::regtype, '3333333333333333333333.3333333333333'::pg_catalog.numeric) where not session_variable.exists('numeric const');
 select session_variable.create_variable('some_name', 'pg_catalog.name'::regtype, 'a_table_or_view_name'::pg_catalog.name) where not session_variable.exists('some_name');
 select session_variable.create_variable('varchar', 'pg_catalog.varchar'::regtype, NULL::pg_catalog.varchar) where not session_variable.exists('varchar');
(7 rows)

select to_char(session_variable.get('some_date', null::date), 'yyyy-mm-dd');               -- fails: has just been dropped
ERROR:  variable or constant 'some_date' does not exists
select session_variable.get('an integer', null::integer);                                  -- fails: has just been dropped
ERROR:  variable or constant 'an integer' does not exists
select session_variable.get('just text', null::text);
 get 
-----
 
(1 row)

select session_variable.get('varchar', null::varchar);
  get   
--------
 23.456
(1 row)

select session_variable.get('numeric const', null::numeric);
                 get                  
--------------------------------------
 3333333333333333333333.3333333333333
(1 row)

create or replace function session_variable.variable_initialisation()
   returns void language plpgsql as $$
begin
	if session_variable.is_executing_variable_initialisation()
	then
	    raise notice 'executing session_variable.variable_initialisation() on behalf of session_variable initialisation code';
	    select session_variable.set('numeric const', 223344);                              -- setting a constant should succeed now  
	else
	    raise notice 'executing session_variable.variable_initialisation() on its own account';
	end if;
end;
$$; 
select session_variable.get('numeric const', null::numeric);
                 get                  
--------------------------------------
 3333333333333333333333.3333333333333
(1 row)

select session_variable.variable_initialisation();
NOTICE:  executing session_variable.variable_initialisation() on its own account
 variable_initialisation 
-------------------------
 
(1 row)

select session_variable.get('numeric const', null::numeric);
                 get                  
--------------------------------------
 3333333333333333333333.3333333333333
(1 row)

select session_variable.init();
NOTICE:  executing session_variable.variable_initialisation() on behalf of session_variable initialisation code
ERROR:  query has no destination for result data
HINT:  If you want to discard the results of a SELECT, use PERFORM instead.
CONTEXT:  PL/pgSQL function session_variable.variable_initialisation() line 6 at SQL statement
SQL statement "select session_variable.variable_initialisation()"
select session_variable.get('numeric const', null::numeric);
  get   
--------
 223344
(1 row)

drop function session_variable.variable_initialisation();
select session_variable.init();
 init 
------
    6
(1 row)

select session_variable.get('numeric const', null::numeric);
                 get                  
--------------------------------------
 3333333333333333333333.3333333333333
(1 row)

select session_variable.drop('just text');
 drop 
------
 t
(1 row)

select session_variable.drop('varchar');
 drop 
------
 t
(1 row)

select session_variable.drop('numeric const');
 drop 
------
 t
(1 row)

select session_variable.drop('initially null');
 drop 
------
 t
(1 row)

select session_variable.drop('does not exist');                                            -- fails: does not exist
ERROR:  variable or constant "does not exist" does not exists
select to_char(session_variable.get('some_date', null::date), 'yyyy-mm-dd');               -- fails: has just been dropped
ERROR:  variable or constant 'some_date' does not exists
select session_variable.get('an integer', null::integer);                                  -- fails: has just been dropped
ERROR:  variable or constant 'an integer' does not exists
select session_variable.get('just text', null::text);                                      -- fails: has just been dropped
ERROR:  variable or constant 'just text' does not exists
select session_variable.get('varchar', null::varchar);                                     -- fails: has just been dropped
ERROR:  variable or constant 'varchar' does not exists
select session_variable.get('numeric const', null::numeric);                               -- fails: has just been dropped
ERROR:  variable or constant 'numeric const' does not exists
-- cleanup
drop schema if exists session_variable cascade;
NOTICE:  drop cascades to extension session_variable
